public/sw.js

```diff
@@ -15,6 +17,120 @@ function hasSameHash(firstUrl, secondUrl) {
 	}
 }
 
+function jsonToArray(obj) {
+	if (typeof obj !== 'object') {
+		return [obj];
+	}
+
+	const keys = [];
+	for (const key in obj) {
+		if (obj.hasOwnProperty(key)) {
+			const children = jsonToArray(obj[key]);
+			children.forEach((child) => {
+				keys.push({ [key]: child });
+			});
+		}
+	}
+
+	return keys;
+}
+
+function arrayToJson(objArray) {
+	let result = {};
+
+	const updateObject = (obj, result) => {
+		const key = Object.keys(obj)[0];
+		result[key] = key in result ? updateObject(obj[key], result[key]) : obj[key];
+		return result;
+	};
+
+	objArray.forEach((obj) => {
+		result = updateObject(obj, result);
+	});
+
+	return result;
+}
+
+async function openDatabase() {
+	return new Promise((resolve, reject) => {
+		const indexedDBOpenRequest = indexedDB.open('post-cache');
+
+		indexedDBOpenRequest.onerror = () => {
+			console.log('IndexedDb error');
+			reject();
+		};
+		indexedDBOpenRequest.onupgradeneeded = function() {
+			this.result.createObjectStore(storeName, { autoIncrement: true, keyPath: 'id' });
+			resolve();
+		};
+
+		indexedDBOpenRequest.onsuccess = function() {
+			db = this.result;
+			resolve();
+		};
+	});
+}
+
+function getObjectStore(readwrite = true) {
+	return db.transaction(storeName, readwrite ? 'readwrite' : 'readonly').objectStore(storeName);
+}
+
+function savePostRequest(url, payload, response) {
+	const keys = jsonToArray(payload);
+	const value = jsonToArray(response);
+	keys.forEach((key, idx) => {
+		getObjectStore().put({
+			url,
+			id: JSON.stringify(key),
+			response: value[idx],
+		});
+	});
+}
+
+function getPostResponse(payload) {
+	return new Promise((resolve) => {
+		let responses = [];
+		const keys = jsonToArray(payload);
+		keys.forEach(async (key, idx) => {
+			await new Promise((next) => {
+				const getRequest = getObjectStore(false).get(JSON.stringify(key));
+				getRequest.onsuccess = function(event) {
+					event.target.result && responses.push(event.target.result.response);
+					next();
+				};
+				getRequest.onerror = function(error) {
+					console.log(error);
+					next();
+				};
+			});
+			if (idx === keys.length - 1) {
+				responses = arrayToJson(responses);
+				const blob = new Blob([JSON.stringify(responses)], { type: 'application/json' });
+				resolve(new Response(blob, { status: 200, statusText: 'OK' }));
+			}
+		});
+	});
+}
+
+function handlePOST(event) {
+	const requestToFetch = event.request.clone();
+	event.respondWith(fetch(requestToFetch, { method: 'POST' })
+		.then(async (response) => {
+			await openDatabase();
```

Comment:
@shubhsherl Instead of every-time opening database here. Can we do like only open when db is null? Will improve performance a bit.

Rest of the things looks perfect. Great work!! 👏 🎉

Reply:
@shubhsherl did you also port this to the upstream PRs?  Just checking..........

Reply:
@ear-dev I'll update in upstream PRs after this one gets merged.